fw/http.c

```diff
@@ -1483,13 +1483,140 @@ tfw_http_nip_req_resched_err(TfwSrvConn *srv_conn, TfwHttpReq *req,
 			  " re-forwarded or re-scheduled");
 }
 
-static inline void
-tfw_http_send_err_resp_nolog(TfwHttpReq *req, int status)
+/**
+ * The request is serviced from cache.
+ * Send the response as is and unrefer its data.
+ */
+static void
+tfw_http_req_cache_service(TfwHttpResp *resp)
 {
+	TfwHttpReq *req = resp->req;
+
+	WARN_ON_ONCE(!list_empty(&req->fwd_list));
+	WARN_ON_ONCE(!list_empty(&req->nip_list));
+
 	if (TFW_MSG_H2(req))
-		tfw_h2_send_err_resp(req, status, false);
+		tfw_h2_resp_fwd(resp);
 	else
-		tfw_h1_send_err_resp(req, status);
+		tfw_http_resp_fwd(resp);
+
+	TFW_INC_STAT_BH(clnt.msgs_fromcache);
+}
+
+/**
+ * Build stale response from the @req->stale_ce and link request with response.
+ * Forward response to client, free previous unsuccessful response from upstream
+ * @hmresp.
+ *
+ * @return true if response successfully forwarded otherwise false.
+ */
+static bool
+__tfw_http_resp_fwd_stale(TfwHttpMsg *hmresp)
+{
+	TfwHttpReq *req = hmresp->req;
+	TfwHttpResp *stale_resp;
+	bool sent = false;
+
+	tfw_stream_unlink_msg(hmresp->stream);
+	/* Unlink response. */
+	req->resp = NULL;
+
+	stale_resp = tfw_cache_build_resp_stale(req);
+	/* For HTTP2 response will not be built if stream already closed. */
+	if (!stale_resp)
+		goto free;
+
+	req->resp->conn = hmresp->conn;
+	hmresp->pair = NULL;
+
+	if (TFW_MSG_H2(req))
+		tfw_h2_req_unlink_stream(req);
+
+	tfw_http_req_cache_service(req->resp);
+	sent = true;
+
+free:
+	tfw_http_msg_free(hmresp);
+
+	return sent;
+}
+
+/**
+ * The same as @__tfw_http_resp_fwd_stale(), but used in case when we don't
+ * have response from upstream.
+ */
+static bool
+__tfw_http_resp_fwd_stale_noresp(TfwHttpReq *req)
+{
+	if (!tfw_cache_build_resp_stale(req))
```

Comment:
If it fails we don't send anything to client. For example when we can't send error responses to client we call `tfw_http_resp_build_error`. We need error injection tests to check Tempesta FW behavior if `tfw_cache_build_resp_stale` and etc. functions fails